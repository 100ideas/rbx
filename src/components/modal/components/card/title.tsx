import { cx } from "emotion";
import React from "react";

import Element from "components/element";
import renderAsExoticComponent from "components/render-as-exotic-component";
import { ModifierProps } from "modifiers";

export type ModalCardTitleModifierProps = Partial<{
  children: React.ReactNode;
  style: {};
}>;

export type ModalCardTitleProps = ModifierProps & ModalCardTitleModifierProps;

const ModalCardTitle = renderAsExoticComponent<ModalCardTitleProps, "p">(
  ({ children, className, ...props }, ref) => (
    <Element {...props} ref={ref} className={cx("modal-card-title", className)}>
      {children}
    </Element>
  ),
  "p",
);
ModalCardTitle.defaultProps = Object.assign(
  {
    children: null,
  },
  ModalCardTitle.defaultProps,
);

export default ModalCardTitle;
