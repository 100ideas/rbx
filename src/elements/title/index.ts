import "./title.module.sass";

export { Title } from "./title";
