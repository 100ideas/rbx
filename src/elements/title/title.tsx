import { cx } from "emotion";
import React from "react";

import { forwardRefAs } from "@/base";
import { ModifierProps, transformModifiers } from "@/modifiers";
import { tuple } from "@/utils";

export const TITLE_SIZES = tuple(1, 2, 3, 4, 5, 6);
export type TitleSizes = (typeof TITLE_SIZES)[number];

export const TITLE_WEIGHTS = tuple("light", "normal", "semibold", "bold");
export type TitleWeights = (typeof TITLE_WEIGHTS)[number];

export type TitleModifierProps = Partial<{
  className: string;
  size: TitleSizes;
  spaced: boolean;
  subtitle: boolean;
  weight: TitleWeights;
}>;

export type TitleProps = ModifierProps & TitleModifierProps;

export const Title = forwardRefAs<TitleProps, "h1">(
  (props, ref) => {
    const { as, size, spaced, subtitle, weight, ...rest } = transformModifiers(
      props,
    );
    rest.className = cx(rest.className, {
      [`is-${size}`]: !!size,
      "is-spaced": spaced && !subtitle,
      subtitle,
      title: !subtitle,
    });
    return React.createElement(as!, { ref, ...rest });
  },
  {
    as: "h1",
    spaced: false,
    subtitle: false,
  },
);
