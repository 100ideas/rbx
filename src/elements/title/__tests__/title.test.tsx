import React from "react";
import renderer from "react-test-renderer";

import { Title } from "../title";

describe("Title component", () => {
  it("should exist", () => {
    expect(Title).toMatchSnapshot();
  });

  it("should have title classname", () => {
    const component = renderer.create(
      <Title>
        Test <a>Give me</a>
      </Title>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should concat classname in props with Bulma classname", () => {
    const component = renderer.create(
      <Title className="other-class this-is-a-test">
        <p>Default</p>
      </Title>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should use inline styles", () => {
    const component = renderer.create(
      <Title style={{ height: 250 }}>
        <p>Default</p>
      </Title>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should be a subtitle with size rendered as P", () => {
    const component = renderer.create(
      <Title<"p"> size={3} subtitle as="p">
        Subtitle
      </Title>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should not be spaced because is subtitle", () => {
    const component = renderer.create(
      <Title<"p"> spaced subtitle as="p">
        Subtitle
      </Title>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should be spaced", () => {
    const component = renderer.create(
      <Title<"p"> spaced as="p">
        Subtitle
      </Title>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
