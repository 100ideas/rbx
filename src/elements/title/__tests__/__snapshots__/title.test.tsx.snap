// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title component should be a subtitle with size rendered as P 1`] = `
<p
  className="is-3 subtitle"
  heading={false}
>
  Subtitle
</p>
`;

exports[`Title component should be spaced 1`] = `
<p
  className="is-spaced title"
  heading={false}
>
  Subtitle
</p>
`;

exports[`Title component should concat classname in props with Bulma classname 1`] = `
<h1
  className="other-class this-is-a-test title"
  heading={false}
>
  <p>
    Default
  </p>
</h1>
`;

exports[`Title component should exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "h1",
    "heading": false,
    "spaced": false,
    "subtitle": false,
  },
  "render": [Function],
}
`;

exports[`Title component should have title classname 1`] = `
<h1
  className="title"
  heading={false}
>
  Test 
  <a>
    Give me
  </a>
</h1>
`;

exports[`Title component should not be spaced because is subtitle 1`] = `
<p
  className="subtitle"
  heading={false}
>
  Subtitle
</p>
`;

exports[`Title component should use inline styles 1`] = `
<h1
  className="title"
  heading={false}
  style={
    Object {
      "height": 250,
    }
  }
>
  <p>
    Default
  </p>
</h1>
`;
